File: deno/day2/ab.ts
const text = await Deno.readTextFile("./input.txt");
let score_A = 0;
let score_B = 0;
for (const line of text.split(/\r?\n/)) {
    const opponent = line.charAt(0).charCodeAt(0) - "A".charCodeAt(0);
    const player = line.charAt(2).charCodeAt(0) - "X".charCodeAt(0);
    score_A += 1 + player + 3 * ((2 * opponent + player + 1) % 3);
    score_B += 1 + player * 3 + (3 + opponent + player - 1) % 3;
}

console.log(score_A)
console.log(score_B)

File: python/day3/ab.py
from pathlib import Path

p = Path(__file__).with_name("input.txt")
with p.open("r") as f:
    all_backpacks = [
        [ord(item) - 96 if ord(item) >= 96 else ord(item) - 38 for item in list(line)]
        for line in f.read().split("\n")
    ]

    score_A = 0

    for backpack in all_backpacks:
        first_half = backpack[: len(backpack) // 2]
        second_half = backpack[len(backpack) // 2 :]
        score_A += sum(set(first_half).intersection(second_half))

    score_B = 0

    for i in range(0, len(all_backpacks), 3):
        first = all_backpacks[i]
        second = all_backpacks[i + 1]
        third = all_backpacks[i + 2]
        score_B += sum(set(first).intersection(set(second).intersection(third)))

    print(score_A)
    assert score_A == 8233

    print(score_B)
    assert score_B == 2821
